api/Updated-Metrics.md

```diff
@@ -0,0 +1,403 @@
+# Updates for Metrics
+
+This is an update for the existing metrics document, which is being left in place for now as a point of comparison.
+
+## Assumed data
+
+In the following we assume that we have variables of the following form defined:
+
+```python
+y_true = [0, 1, 0, 0, 1, 1, ...]
+y_pred = [1, 0, 0, 1, 0, 1, ...] # Content can be different for other metrics (see below)
+A_1 = [ 'C', 'B', 'B', 'C', ...]
+A_2 = [ 'M', 'N', 'N', 'P', ...]
+A = pd.DataFrame(np.transpose([A_1, A_2]), columns=['SF 1', 'SF 2'])
+
+weights = [ 1, 2, 3, 2, 2, 1, ...]
+```
+
+We actually seek to be very agnostic as to the contents of the `y_true` and `y_pred` arrays; the meaning is imposed on them by the underlying metrics.
+Here we have shown binary values for a simple classification problem, but they could be floating point values from a regression, or even collections of classes and associated probabilities.
+
+## Basic Calls
+
+### Existing Syntax
+
+Our basic method is `group_summary()`
+
+```python
+>>> result = flm.group_summary(skm.accuracy_score,
+                               y_true, y_pred,
+                               sensitive_features=A_1)
+>>> print(result)
+{'overall': 0.4, 'by_group': {'B': 0.6536, 'C': 0.213}}
+>>> print(type(result))
+<class 'sklearn.utils.Bunch'>
+```
+The `Bunch` is an object which can be accessed in two ways - either as a dictionary - `result['overall']` -  or via properties named by the dictionary keys - `result.overall`.
+Note that the `by_group` key accesses another `Bunch`.
+
+We allow for sample weights (and other arguments which require slicing) via `indexed_params`, and passing through other arguments to the underlying metric function (in this case, `normalize`):
+```python
+>>> flm.group_summary(skm.accuracy_score,
+                      y_true, y_pred,
+                      sensitive_features=A_1,
+                      indexed_params=['sample_weight'],
+                      sample_weight=weights, normalize=False)
+{'overall': 20, 'by_group': {'B': 60, 'C': 21}}
+```
+
+We also provide some wrappers for common metrics from SciKit-Learn:
+```python
+>>> flm.accuracy_score_group_summary(y_true, y_pred, 
+                                     sensitive_features=A_1)
+{'overall': 0.4, 'by_group': {'B': 0.6536, 'C': 0.213}}
+```
+
+### Proposed Change
+
+We propose to introduce a new object, the `GroupedMetric` (name discussion below).
+Users will compute metrics by passing arguments into the constructor:
+```python
+>>> metrics = GroupedMetrics(skm.accuracy_score,
+                             y_true, y_pred,
+                             sensitive_features=A_1)
+<class 'GroupedMetric'>
+>>> metrics.overall
+          accuracy_score
+overall   0.4
+>>> metrics.by_group
+   accuracy_score
+B             0.6536
+C             0.213
+```
+Both properties are now Pandas DataFrames, with the column name set to the `__name__` property of the given metric function.
```

Comment:
As I start trying to write the `GroupedMetric` I'm finding this is potentially an issue. If I make both properties into DataFrames, then you end up having to write code like:
```python
assert target.overall['recall_score'][0] == expected_overall
assert target.by_group['recall_score']['A'] == expected_A
```
This is obviously going to get quite tedious - however, it will always be consistent.

The problem is most acute for the `overall` property. In the most common case, users will be expecting a scalar. The exceptions are:
- If the user has specified `conditional_features=` then the `overall` property naturally become a Series, with the index set to the unique values (or combinations of) the conditional feature(s)
- If the user has specified multiple metrics, then `overall` will be a DataFrame, with columns corresponding to the different metrics (and multiple rows with `conditional_features=`)

Similarly, `by_group` can be a Series unless we have more than one metric _or_  we have conditional features.

So we have a choice:
1. Consistently use DataFrames, making the common case more verbose
1. Change the returned data types based on the supplied data, accepting increased internal complexity, and also making any consumers of the library adjust their expectations according to their supplied data

Reply:
we could have the output of the object methods as a DF and the output of the helper functions a scalar?